### PR TITLE
Refining status when Condition failed

### PR DESCRIPTION
**Pull request by vincent-pli**
Introduce new state: `SucceededWithConditionCheckFailed`, indicates the scenario `pipeline` succeeded but with `Condition` checking failed. 
fix issue: #1692 

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
Status messages for completed and running PipelineRuns have been updated to include the count of succeeded, skipped, and incomplete Tasks. Additionally a new status Reason, "Completed", has been introduced for finished PipelineRuns that have skipped Tasks due to Condition check failures.

### CODE REVIEW

**Comment by afrittoli** [comment]
Thank you for this patch, and sorry I'm a bit late in the discussion.
While it could be interesting to signal that a pipeline did not execute any task, I think `SucceededWithConditionCheckFailed` does not express this information. I could have a pipeline with several conditions, some passed, some failed, which could be successful or failed. Perhaps `SucceededWithNoTaskRuns` would represent better the situation? 
Note that it is already possible to know that no TaskRun was executed, so this would only make it slightly easier to discover. I'm not entirely convinced it is needed though. Do you have specific use cases in mind to explain why this is required?

**Comment by vincent-pli** [comment]
@afrittoli thanks for comments :)
The `SucceededWithConditionCheckFailed ` indicate that there is at least one condition checking failed in the whole `pipelinerun`. Just a reminder for user to hint there is something abnormal occurred to avoid misleading.

`SucceededWithNoTaskRuns ` express there is no `task` really executed since `condition` checking failed.
I think just a reminder is enough for end user.

**Comment by dibyom** [comment]
So, we have a `Reason` and a `Message` field. I think what could be confusing for the user is the `Message` saying `All Tasks have completed executing` when no tasks have been executed. So, my two cents would be to use a different message like `No Tasks left to execute` or something and keep the Reason the same as before.

**Comment by ImJasonH** [comment]
I think @dibyom is right that we can/should use the `Message` field better, whether or not these different states end up having different `Reason`s.

I like the `Reason` being `no more tasks to run` because that's both true, and meaningfully vague. :smile: 

The `Message` then could be `completed 10 tasks, 3 were not executed due to conditions` (or something along those lines). This could even be useful when the PipelineRun is ongoing (`completed 6 tasks, running 2 tasks, 1 skipped due to conditions`). This of course is something a user should also be able to determine from taskrun statuses, but it's nice to have a human-readable summary, and it should be fairly straightforward to generate it.

**Comment by vincent-pli** [comment]
Could we just use `Completed` for `reason` rather than `Succeeded`
then we have message for both `Complete` and `Running`: 
`Completed 6 tasks, not start/running 2 tasks, skipped 1 tasks`
@afrittoli @dibyom @ImJasonH

**Comment by dibyom** [comment]
> Could we just use Completed for reason rather than Succeeded
SGTM especially since `condition.type: Succeeded` already. What do others think -- @ImJasonH @vdemeester @bobcatfish ?

**Comment by sbwsg** [comment]
Hey @vincent-pli do you intend to continue working on this or should we close for now?  For what it's worth it sounds like we have consensus on the desired implementation:

> Could we just use Completed for reason rather than Succeeded
then we have message for both Complete and Running:
Completed 6 tasks, not start/running 2 tasks, skipped 1 tasks

SGTM2

**Comment by vincent-pli** [comment]
@sbwsg 
Sure, I will work on it soon.

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 1]
```diff
@@ -436,7 +450,7 @@ func GetPipelineConditionStatus(pr *v1alpha1.PipelineRun, state PipelineRunState
 		Type:    apis.ConditionSucceeded,
 		Status:  corev1.ConditionUnknown,
 		Reason:  ReasonRunning,
-		Message: "Not all Tasks in the Pipeline have finished executing",
+		Message: fmt.Sprintf("Completed %d tasks, not start/running %d tasks, skipped %d tasks", len(successOrSkipTasks)-skipTasks, len(allTasks)-len(successOrSkipTasks), skipTasks),
```
Comment (sbwsg) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go]: Suggest "Tasks Completed: %d, Incomplete: %d, Skipped: %d" here to make it a little shorter / quicker to read.

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 2]
```diff
@@ -411,22 +415,32 @@ func GetPipelineConditionStatus(pr *v1alpha1.PipelineRun, state PipelineRunState
...
+			reason = ReasonCompleted
+		}
+
 		return &apis.Condition{
 			Type:    apis.ConditionSucceeded,
 			Status:  corev1.ConditionTrue,
-			Reason:  ReasonSucceeded,
-			Message: "All Tasks have completed executing",
+			Reason:  reason,
+			Message: fmt.Sprintf("Completed %d tasks, skipped %d tasks", len(successOrSkipTasks)-skipTasks, skipTasks),
```
Comment (sbwsg) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go]: Suggest "Tasks Completed: %d, Skipped: %d" here too.

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go (hunk 1 above)
Comment (vincent-pli) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go]: @sbwsg  Done, thanks

**Comment by vincent-pli** [comment]
@sbwsg For the release, I think I agree with you.

**Comment by sbwsg** [comment]
>  For the release, I think I agree with you.

OK cool, I suggest updating the Release Notes on this PR's description to be something like:

```
Status messages for completed and running PipelineRuns have been updated to include the count of succeeded, skipped, and incomplete Tasks. Additionally a new status Reason, "Completed", has been introduced for finished PipelineRuns that have skipped Tasks due to Condition check failures.
```

Then I think this is ready to merge!

**Comment by vincent-pli** [comment]
@sbwsg 
Done, thanks

**Pull request merged**